arch/riscv/kernel/setup.c

```diff
@@ -267,6 +268,7 @@ static void __init parse_dtb(void)
 void __init setup_arch(char **cmdline_p)
 {
 	parse_dtb();
+	promote_to_cove_guest();
```

Comment:
Can you invoke it after sbi_init and riscv_cove_sbi_init so that this can be called only if is_cove_guest is true

Reply:
The promote call will actually move to different place, i.e., head.S in `_start_kernel`. Like this, the kernel requests the promotion before it touches the stack or heap, so there is no runtime state associated with the execution. Consequently the TVM's memory at the time of promotion equals the linux kernel image, so integrity measurements are easy to calculate. 

The approach presented in this patch was intended to match what OpenPower PEF did. But in such a case we would need to do a few tricks that would unnecessarily complicate the implementation.

I will present details on local attestation during RV CoVE TG on November 5th.

Reply:
The commit #fb1b6080a118b70305be6bb1e0243a5e1b119e05 implements the simpler approach to VM promotion, which I mentioned above.

Reply:
If you want to convert so early, why not just do it from the VMM ? The VMM can create a single step or multi-step TVM. It should be even simpler ?

Reply:
Another thing is that for embedded systems we need local attestation, i.e., a way to attest to TVM's integrity and authenticity without network access. To do so, we must give the TSM a proof that the VM being promoted is authorized to access certain secrets. Following what we did for Openpower PEF, we embed the proof (defined in the CoVE spec as TVM Attestation Payload (TAP)) inside the VM's image. Kernel knows the location of TAP and exposes it to TSM when making the promote call. The problematic thing about doing the promotion at the VMM-level, a part from the increased TCB, is that we would need to make VMM aware of TAP and explicitly pass TAP to the VMM.

Reply:
Deployment model 3 is primarily aimed at embedded systems although it can also be used by cloud or other systems. Putting the execution in the kernel as proposed is a minimal code approach. This approach will fit better into the development cycles for embedded systems. The tooling to generate the Attestation Payload will be simpler. There is no need to actually run the VM prior to generating the payload. If the VMM does not support confidential computing the call will fail which is not a problem because no secrets will be exposed because VM will not have access to the key that unlocks its disk. This approach gives the VM creator the ability to lock their TVM to a specific set of hardware. To do this requires that each platform have unique information, such as a serial number, that is incorporated into the attestation. The TVM can be locked to a single platform or a set of platforms, depending on the hardware environment.  This is also important for embedded systems.

Reply:
Just to clarify: I was not suggesting to trap any VM's execution at any certain stage and promote it to TVM.
I was suggesting to start the TVM in the single step boot mode at the beginning itself. No change in promote_to_tvm API.

Instead of this path
VM->(promote_to_tvm)->TSM traps and forward ->Host->promote_to_tvm (COVH)->TSM

We can just simply start the TVM. Are there any issue with this execution model ?
VMM->host (start TVM in single_step_mode)

The flexibility you mentioned still exists for anybody to implement in a different way. I was mainly concerned about the odd boot flow for a Linux based TVM. My earlier assumption was a Linux VM can be promoted to TVM at any point during the boot. If it is supposed to promoted right after the boot, VMM may as well tell the host to start a single step TVM. I understand that the firmware needs to be part of the TCB in this case but I am not sure if we need that for this use case.

We will anyways have have VMM modification for CoVE for other models. So adding another ABI for single step booting shouldn't be a issue.

Reply:
@atishp04 thanks for the explanation.

The current path is different that what you described. The VM traps into Host, which reflects the call to TSM:
VM->(promote_to_tvm)->Host->promote_to_tvm (COVH)->TSM

If I understood correctly, you propose that VMM+Host do the promotion of a Linux VM that hasn't executed yet, right?  
VMM->(new promote ABI)->Host->promote_to_tvm (COVH)->TSM

Your motivation to shift promotion call from VM to VMM is that it sounds odd to call promote_to_tvm ABI early in Linux kernel and the same effect could be achieved by doing promotion before the VM execution, right? I agree with you that we can call promote_to_tvm before kernel execution (what you propose) and we can do it during early kernel boot (current proposal) or later during runtime (this will be more like a snapshots of a running system). 

In Linux kernel's powerpc (Openpower PEF) we promote TVMs the same way as proposed now for risc-v, so within first instructions after entering the Linux kernel code. Starting a TVM in a single-step-mode would not support use cases with firmware that is not intended to be in TCB and would require modification of every VMM, something we can discuss today during the call. Do you know what are the plans for VMM modifications? Do you know which VMMs are expected to get support for CoVE (qemu, kvmtool, firecracker)?

Some other question I have is how could we make a VMM aware of TAP? Should we pass it as an extra argument or should VMM somehow extract TAP address from the VM image? Current approach embeds TAP inside the Linux kernel image and Linux kernel is aware about the TAP's address so it can expose it during the promote call.